src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,311 @@
+    """ Generates the cbsd,ppa and esc_sensor dump file and dumps the FAD file in specified path
+
+    Args:
+       cbsd_records: list of cbsdData objects that includes one or more grants object and FAD
+       grant_records_list: list of muliple grant records for all cbsd devices
+         For an example : muliple grants [ [grant1, grant2], [grant1, grant2] ] associated with each
+         cbsd records [ cbsd1, cbsd2 ]
+       ppa_records:  list of ppa_record objects used to generate ppa_dump file and FAD
+       esc_sensor_records: list of esc_sensor_records objects used to generate
+                            esc_sensor_records and FAD
+
+    Returns:
+       It returns None and successfully generates the dumps in specified file path
+    """
+
+    def getCbsdReferenceId(fcc_id, serial_number):
+      """creates encoded cbsd_reference_id using sha1 with fcc_id and serial number of cbsd
+      """
+      return str(fcc_id + '/' + str(hashlib.sha1(serial_number).hexdigest())).encode('utf-8')
+
+    def writeDumpFile(file_path, data):
+      """ Write JSON data to specified file """
+      file_handler = open(file_path, "w")
+      file_handler.writelines(json.dumps(data, sort_keys=True, indent=4))
+      file_handler.close()
+
+    def cleanDumpFiles():
+      """ Clean existing dumpfile if any """
+      map(os.remove, (os.path.join(self.getDumpFilePath(), file)
+                      for file in os.listdir(self.getDumpFilePath())))
+
+    def createFadRecord(url, record):
+      """ Creates the files object that includes url,checksum,size,version and recordtype of record
+          required to be packed in FAD
+      """
+      fad_record = {}
+      fad_record['url'] = url
+      fad_record['checksum'] = hashlib.sha1(json.dumps(record)).hexdigest()
+      fad_record['size'] = sys.getsizeof(record)
+      fad_record['version'] = self.getSasThUrl().split('/')[-1]
+      if 'cbsd' in url:
+        fad_record['recordType'] = "cbsd"
+      elif 'zone' in url:
+        fad_record['recordType'] = "zone"
+      elif 'esc_sensor' in url:
+        fad_record['recordType'] = "esc_sensor"
+      else:
+        pass
+      return fad_record
+
+    def createFadObject(all_activity_dump):
+      """Creates the FAD object format and embed the active dump file object of different recordtype
+      """
+      full_activity_dump = {}
+      full_activity_dump['files'] = all_activity_dump
+      end_time = datetime.now()
+      end_time = time.strftime(end_time.replace(microsecond=0).isoformat()) + 'Z'
+      full_activity_dump['generationDateTime'] = end_time
+      full_activity_dump['description'] = 'Load {} FAD to SAS UUT'.format(self.name)
+      return full_activity_dump
+
+    fad_record_list = []
+
+    # Before Generate FAD record cleanup if any dump file
+    logging.debug('Clean dump files if any exist already')
+    cleanDumpFiles()
+    for index, (cbsd_record, grant_records) in enumerate(zip(cbsd_records, grant_records_list)):
```

Comment:
The   grant_records_list contains a list of grants for each cabs.          We have redesigned the SAS test harness to address this comment. The   configuration file now contains the cabs records, ppa record and esc record.   Utility functions are used to create cabs, ppa and esc records conforming to   the SAS-SAS interface format from the    existing record types in the testdata folder. The related functions   are written as static methods so that they can be invoked within the generate   default config method before creating an object for SAS test harness.